src/services/meta_transaction_service.ts

```diff
@@ -0,0 +1,196 @@
+        const { takerAddress, sellAmount, buyAmount, swapQuote, price } = await this.calculateMetaTransactionPriceAsync(
+            params,
+        );
+
+        const gasPrice = swapQuote.gasPrice;
+        const attributedSwapQuote = serviceUtils.attributeSwapQuoteOrders(swapQuote);
+        const orders = attributedSwapQuote.orders;
+        const signatures = orders.map(order => order.signature);
+
+        const zeroExTransaction = this._generateZeroExTransaction(
+            orders,
+            sellAmount,
+            buyAmount,
+            signatures,
+            takerAddress,
+            gasPrice,
+        );
+
+        // use the DevUtils contract to generate the transaction hash
+        const devUtils = new DevUtilsContract(this._contractWrappers.contractAddresses.devUtils, this._provider);
+        const zeroExTransactionHash = await devUtils
+            .getTransactionHash(
+                zeroExTransaction,
+                new BigNumber(CHAIN_ID),
+                this._contractWrappers.contractAddresses.exchange,
+            )
+            .callAsync();
+
+        const makerAssetAmount = swapQuote.bestCaseQuoteInfo.makerAssetAmount;
+        const totalTakerAssetAmount = swapQuote.bestCaseQuoteInfo.totalTakerAssetAmount;
+        const apiMetaTransactionQuote: GetMetaTransactionQuoteResponse = {
+            price,
+            zeroExTransactionHash,
+            zeroExTransaction,
+            buyAmount: makerAssetAmount,
+            sellAmount: totalTakerAssetAmount,
+            orders: serviceUtils.cleanSignedOrderFields(orders),
+        };
+        return apiMetaTransactionQuote;
+    }
+    private _generateZeroExTransaction(
+        orders: SignedOrder[],
+        sellAmount: BigNumber | undefined,
+        buyAmount: BigNumber | undefined,
+        signatures: string[],
+        takerAddress: string,
+        gasPrice: BigNumber,
+    ): ZeroExTransaction {
+        // generate txData for marketSellOrdersFillOrKill or marketBuyOrdersFillOrKill
+        let txData;
+        if (sellAmount !== undefined) {
+            txData = this._contractWrappers.exchange
+                .marketSellOrdersFillOrKill(orders, sellAmount, signatures)
+                .getABIEncodedTransactionData();
+        } else if (buyAmount !== undefined) {
+            txData = this._contractWrappers.exchange
+                .marketBuyOrdersFillOrKill(orders, buyAmount, signatures)
+                .getABIEncodedTransactionData();
+        } else {
+            throw new Error('sellAmount or buyAmount required');
+        }
+
+        // generate the zeroExTransaction object
+        const expirationTimeSeconds = new BigNumber(Date.now() + TEN_MINUTES_MS)
+            .div(ONE_SECOND_MS)
+            .integerValue(BigNumber.ROUND_CEIL);
+        const zeroExTransaction: ZeroExTransaction = {
+            data: txData,
+            salt: generatePseudoRandomSalt(),
+            signerAddress: takerAddress,
+            gasPrice,
```

Comment:
TODO we should round this up to the nearest integer of full gwei so they're bucketed a bit closer